Documentation changes for non-ISO calendars (#1321)
tc39/proposal-temporal: docs/calendar.md

> **NOTE:** The Temporal polyfill currently only includes the ISO 8601 calendar, but the Temporal proposal will eventually provide all the calendars supported by Intl.
> See [issue #541](https://github.com/tc39/proposal-temporal/issues/541).

> ### When to use `Temporal.Calendar`

Reviewer: Thanks. This has been annoying me for a while, but I keep forgetting to fix it. :rofl:

date.withCalendar(calendar).add({ months: 1 });
```

### Invariants Across Calendars

> The following "invariants" (statements that are always true) hold for all built-in calendars, and SHOULD hold for properly-authored custom calendars:

Author: @sffc, I'll revise to this:

```suggestion
The following "invariants" (statements that are always true) hold for all built-in calendars, and should also hold for any properly-authored custom calendar that supports years, months, and days units:
```

> - `year`, `month`, `day` are always integers that increase as time goes forward
> - `month` and `day` are always positive integers

Reviewer: This might need to be phrased a bit differently for `month` and `day`, how about this?
```suggestion
- `year` is always an integer that increases as time goes forward
- `month` and `day` are always positive integers that increase as time goes forward, except they reset at the boundary of a year or month, respectively
```

- `date.month === 1` during the first month of any year
- `date.month === date.monthsInYear` during the last month of any year
- `month` is always continuous (no gaps)
- Any date can be serialized to an object using only four properties: `{ year, month, day, calendar }`

> ### Writing Cross-Calendar Code

Reviewer: I really like this section. I think it's fine here for now. It makes the page a bit long so it could be moved to its own page in the future, but I wouldn't delay the PR on that.

> - Any date can be serialized to an object using only four properties: `{ year, month, day, calendar }`

Reviewer 2: I wouldn't make this claim, because we can't guarantee this to be true for custom calendars.  The ISO fields are the things guaranteed to round-trip.

Reviewer: I think it's claimed to be the case for builtin calendars, and recommended for custom calendars - see the language on line 41.

Reviewer 2: I don't know though if I'd go so far to claim this invariant SHOULD hold for custom calendars.  This works so long as custom calendars use something that roughly maps to years, months, and days.  But a custom calendar could use wholly different units.  We often use shamrocks, rainbows, and leprechauns as an example.

Author: OK I checked in new text above.

Here are best practices for writing code that will work regardless of the calendar used:

> - Validate or coerce the calendar of all external input. If your code receives date data from an external source, you should validate that its calendar is what you expect or coerce it to a known calendar via the `withCalendar` method. Otherwise, a sophisticated attacker could change the behavior of your app or introduce security or performance issues by introducing an unexpected calendar.

Reviewer: I think this might need some rephrasing. It's not likely that code would receive a Temporal object in an _unknown_ calendar, unless they had a dependency on a userland calendar somewhere in their stack. Probably what we want to advise people to guard against is _unexpected_ calendars.

How about this?
```suggestion
- Validate or coerce the calendar of all external input. If your code receives a Temporal object from an external source, you should check that its calendar is what you expect, and if you are not prepared to handle other calendars, convert it to the ISO calendar using `withCalendar('iso8601')`. Otherwise, you may end up with unexpected behavior in your app or introduce security or performance issues by introducing an unexpected calendar.
```

- Use `compare` methods (e.g. `Temporal.PlainDate.compare(date1, '2000-01-01')`) instead of manually comparing individual properties (e.g. `date.year > 2000`) whose meaning may vary across calendars.
- Never compare field values in different calendars. A `month` or `year` in one calendar is unrelated to the same property values in another calendar. If dates in different calendars must be compared, use `compare`.
- When comparing dates for equality that might be in different calendars, convert them both to the same calendar using `withCalendar`. The same ISO date in different calendars will return `false` from the `equals` method and will return a non-zero value from `compare` because the calendars are not equal.

> - When looping through all months in a year, use `monthsPerYear` as the upper bound instead of assuming that every year has 12 months.

Reviewer: ```suggestion
- When looping through all months in a year, use `monthsInYear` as the upper bound instead of assuming that every year has 12 months.
```

- Don't assume that `date.month===12` is the last month of the year. Instead, use `date.month===date.monthsInYear`.
- Use `until` or `since` to count years, months, or days between dates. Manually calculating differences (e.g. `Math.floor(months/12)`) will fail for some calendars.

> - Use `daysPerMonth` instead of assuming that each month has the same number of days in every year.

Reviewer: ```suggestion
- Use `daysInMonth` instead of assuming that each month has the same number of days in every year.
```

This method does not need to be called directly except in specialized code.
It is called indirectly when using the `from()` static methods and `with()` methods of `Temporal.PlainDateTime`, `Temporal.PlainDate`, and `Temporal.PlainYearMonth`.

Custom calendars should override this method if they accept fields in `from()` or `with()` other than the standard set of built-in calendar fields: `year`, `month`, `monthCode`, and `day`.

> The input array contains the field names that are necessary for a particular operation (for example, `'monthCode'`, and `'day'` for `Temporal.PlainMonthDay.prototype.with()`).

Reviewer: ```suggestion
The input array contains the field names that are necessary for a particular operation (for example, `'monthCode'` and `'day'` for `Temporal.PlainMonthDay.prototype.with()`).
```

<!-- prettier-ignore-start -->
```js

> // In built-in calendars, this method just makes a copy of the input array

Reviewer: ```suggestion
// In the ISO calendar, this method just makes a copy of the input array
```
